Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,658 @@
+
+theorem b₈_of_char_two : W.b₈ = W.a₄ ^ 2 := by
+  rw [self.b₈]
+  linear_combination -W.a₄ ^ 2 * CharP.cast_eq_zero R 2
+
+theorem c₄_of_char_two : W.c₄ = 0 := by
+  rw [self.c₄]
+  linear_combination -24 * W.a₄ * CharP.cast_eq_zero R 2
+
+theorem c₆_of_char_two : W.c₆ = 0 := by
+  rw [self.c₆]
+  linear_combination -108 * W.b₆ * CharP.cast_eq_zero R 2
+
+theorem Δ_of_char_two : W.Δ = W.a₃ ^ 4 := by
+  rw [self.Δ, W.b₆_of_char_two]
+  linear_combination (-32 * W.a₄ ^ 3 - 14 * W.a₃ ^ 4) * CharP.cast_eq_zero R 2
+
+end IsCharTwoJZeroNF
+
+namespace IsCharTwoJZeroNF
+
+variable {E} (self : E.IsCharTwoJZeroNF)
+include self
+
+theorem j_of_char_two [CharP F 2] : E.j = 0 := by
+  simp [EllipticCurve.j, self.c₄_of_char_two]
+
+end IsCharTwoJZeroNF
+
+/-! ### Normal forms of characteristic = 2 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic = 2, if it is
+$Y^2 + XY = X^3 + a_2X^2 + a_6$ (`WeierstrassCurve.IsCharTwoJNeZeroNF`) or
+$Y^2 + a_3Y = X^3 + a_4X + a_6$ (`WeierstrassCurve.IsCharTwoJZeroNF`). -/
+def IsCharTwoNF : Prop := W.IsCharTwoJNeZeroNF ∨ W.IsCharTwoJZeroNF
+
+section VariableChange
+
+variable [CharP R 2] [CharP F 2]
+
+/-- For a `WeierstrassCurve` defined over a ring of characteristic = 2,
+there is an explicit change of variables, which changes it to $Y^2 + a_3Y = X^3 + a_4X + a_6$
+(`WeierstrassCurve.IsCharTwoJZeroNF`) if its j = 0. -/
+def toCharTwoJZeroNF : VariableChange R := ⟨1, W.a₂, 0, 0⟩
+
+/-- For a `WeierstrassCurve` defined over a field of characteristic = 2,
+there is an explicit change of variables, which changes it to $Y^2 + XY = X^3 + a_2X^2 + a_6$
+(`WeierstrassCurve.IsCharTwoJNeZeroNF`) if its j ≠ 0. -/
+def toCharTwoJNeZeroNF (W : WeierstrassCurve F) (ha₁ : W.a₁ ≠ 0) : VariableChange F :=
+  ⟨Units.mk0 _ ha₁, W.a₃ / W.a₁, 0, (W.a₁ ^ 2 * W.a₄ + W.a₃ ^ 2) / W.a₁ ^ 3⟩
+
+theorem toCharTwoJZeroNF_spec (ha₁ : W.a₁ = 0) :
+    (W.variableChange W.toCharTwoJZeroNF).IsCharTwoJZeroNF := by
+  constructor
+  · simp [toCharTwoJZeroNF, ha₁]
+  · simp [toCharTwoJZeroNF, show W.a₂ + 3 * W.a₂ = 0 by
+      linear_combination 2 * W.a₂ * CharP.cast_eq_zero R 2]
+
+section
+
+variable (W : WeierstrassCurve F)
+
+theorem toCharTwoJNeZeroNF_spec (ha₁ : W.a₁ ≠ 0) :
+    (W.variableChange (W.toCharTwoJNeZeroNF ha₁)).IsCharTwoJNeZeroNF := by
+  constructor
+  · simp [toCharTwoJNeZeroNF, ha₁]
+  · field_simp [toCharTwoJNeZeroNF]
+    linear_combination (W.a₃ * W.a₁ ^ 3 + W.a₁ ^ 2 * W.a₄ + W.a₃ ^ 2) * CharP.cast_eq_zero F 2
+  · field_simp [toCharTwoJNeZeroNF]
+    linear_combination (W.a₁ ^ 4 * W.a₃ ^ 2 + W.a₁ ^ 5 * W.a₃ * W.a₂) * CharP.cast_eq_zero F 2
+
+/-- For a `WeierstrassCurve` defined over a field of characteristic = 2,
+there is an explicit change of variables of it to `WeierstrassCurve.IsCharTwoNF`, that is,
+$Y^2 + XY = X^3 + a_2X^2 + a_6$ (`WeierstrassCurve.IsCharTwoJNeZeroNF`) or
+$Y^2 + a_3Y = X^3 + a_4X + a_6$ (`WeierstrassCurve.IsCharTwoJZeroNF`). -/
+def toCharTwoNF [DecidableEq F] : VariableChange F :=
```

Comment:
I saw that you made a post on Zulip if this should be `[DecidableEq F]` or `open scoped Classical in`, and you decided on the former, because you want them to be computable if `F` is a computable field. This is more of a question for my sanity: should I do the same for things like point addition if I could?

Reply:
> This is more of a question for my sanity: should I do the same for things like point addition if I could?

It's up to you. Personally I think `DecidableEq` is better. (Perhaps in some day Lean can be a number theory CAS with these algorithms.)

Reply:
Presumably you can't do `[DecidableEq]` for `exists_variableChange_isCharTwoNF`?

Reply:
> Presumably you can't do `[DecidableEq]` for `exists_variableChange_isCharTwoNF`?

It's also possible. Will change it.